tests/bluetooth/bsim_bt/bsim_test_gatt_settings/src/gatt_utils.c

```diff
@@ -0,0 +1,368 @@
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/sys/__assert.h>
+
+/* Custom Service Variables */
+static struct bt_uuid_128 test_svc_uuid = BT_UUID_INIT_128(
+	0xf0, 0xde, 0xbc, 0x9a, 0x78, 0x56, 0x34, 0x12,
+	0x78, 0x56, 0x34, 0x12, 0x78, 0x56, 0x34, 0x12);
+
+static struct bt_uuid_128 test_svc_uuid_2 = BT_UUID_INIT_128(
+	0xf1, 0xdd, 0xbc, 0x9a, 0x78, 0x56, 0x34, 0x12,
+	0x78, 0x56, 0x34, 0x12, 0x78, 0x56, 0x34, 0x12);
+
+static struct bt_uuid_128 test_chrc_uuid = BT_UUID_INIT_128(
+	0xf2, 0xde, 0xbc, 0x9a, 0x78, 0x56, 0x34, 0x12,
+	0x78, 0x56, 0x34, 0x12, 0x78, 0x56, 0x34, 0x12);
+
+static uint8_t test_value[] = { 'T', 'e', 's', 't', '\0' };
+
+DEFINE_FLAG(flag_client_read);
+
+static ssize_t read_test(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			void *buf, uint16_t len, uint16_t offset)
+{
+	const char *value = attr->user_data;
+
+	printk("Client has read from test char\n");
+	SET_FLAG(flag_client_read);
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, value,
+				 strlen(value));
+}
+
+void wait_for_client_read(void)
+{
+	WAIT_FOR_FLAG(flag_client_read);
+}
+
+static ssize_t write_test(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			 const void *buf, uint16_t len, uint16_t offset,
+			 uint8_t flags)
+{
+	uint8_t *value = attr->user_data;
+
+	printk("Client has written to test char\n");
+
+	if (offset + len > sizeof(test_value)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	memcpy(value + offset, buf, len);
+
+	return len;
+}
+
+static struct bt_gatt_attr test_attrs[] = {
+	/* Vendor Primary Service Declaration */
+	BT_GATT_PRIMARY_SERVICE(&test_svc_uuid),
+
+	BT_GATT_CHARACTERISTIC(&test_chrc_uuid.uuid,
+			       BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE,
+			       BT_GATT_PERM_READ_ENCRYPT |
+			       BT_GATT_PERM_WRITE_ENCRYPT,
+			       read_test, write_test, test_value),
+};
+
+static struct bt_gatt_attr test_attrs_2[] = {
+	/* Vendor Primary Service Declaration */
+	BT_GATT_PRIMARY_SERVICE(&test_svc_uuid_2),
+
+	BT_GATT_CHARACTERISTIC(&test_chrc_uuid.uuid,
+			       BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE,
+			       BT_GATT_PERM_READ_ENCRYPT |
+			       BT_GATT_PERM_WRITE_ENCRYPT,
+			       read_test, write_test, test_value),
+};
+
+static struct bt_gatt_service test_svc = BT_GATT_SERVICE(test_attrs);
+static struct bt_gatt_service test_svc_2 = BT_GATT_SERVICE(test_attrs_2);
+
+void gatt_register_service_1(void)
+{
+	int err = bt_gatt_service_register(&test_svc);
+
+	__ASSERT(!err, "Failed to register GATT service (err %d)\n", err);
+}
+
+void gatt_register_service_2(void)
+{
+	/* This service is only used to trigger a GATT DB change.
+	 * No reads or writes will be attempted.
+	 */
+	int err = bt_gatt_service_register(&test_svc_2);
+
+	__ASSERT(!err, "Failed to register GATT service (err %d)\n", err);
+}
+
+/* We need to discover:
+ * - Dynamic service
+ * - Client Features (to set robust caching)
+ * - DB hash handle (to get out of the change-unaware state)
```

Comment:
I dont see the DB hash is read anywhere? Also a thing to note, the DB hash shall use read by uuid:
> In order to read the value of this characteristic the client shall always use the
GATT Read Using Characteristic UUID sub-procedure. The Starting Handle
should be set to 0x0001 and the Ending Handle should be set to 0xFFFF.

Not sure it is worth caring about in a test.

Reply:
I was reading it in a previous version. I'll remove it.